File: src/python/nispor/hsr.py
# SPDX-License-Identifier: Apache-2.0

from .base_iface import NisporBaseIface


class NisporHsr(NisporBaseIface):
    def __init__(self, info):
        super().__init__(info)
        self._hsr_info = self._info.get("hsr", {})

    @property
    def port1(self):
        return self._hsr_info.get("port1")

    @property
    def port2(self):
        return self._hsr_info.get("port2")

    @property
    def interlink(self):
        return self._hsr_info.get("interlink")

    @property
    def supervision_addr(self):
        return self._hsr_info.get("supervision_addr")

    @property
    def seq_nr(self):
        return self._hsr_info.get("seq_nr")

    @property
    def multicast_spec(self):
        return self._hsr_info.get("multicast_spec")

    @property
    def version(self):
        return self._hsr_info.get("version")

    @property
    def protocol(self):
        return self._hsr_info.get("protocol")

File: src/lib/integ_tests/hsr.rs
// SPDX-License-Identifier: Apache-2.0

use std::panic;

use pretty_assertions::assert_eq;

use super::utils::assert_value_match;
use crate::NetState;

const IFACE_NAME: &str = "hsr0";

// seq_nr has been excluded as it is non-deterministic
const EXPECTED_HSR_INFO: &str = r#"---
name: hsr0
type: hsr
hsr:
  port1: eth1
  port2: eth2
  supervision-addr: 01:15:4e:00:01:2d
  multicast-spec: 0
  version: 0
  protocol: prp"#;

const EXPECTED_HSR_INTERLINK_INFO: &str = r#"---
name: hsr0
type: hsr
hsr:
  port1: eth1
  port2: eth2
  interlink: eth3
  supervision-addr: 01:15:4e:00:01:2d
  multicast-spec: 0
  version: 0
  protocol: hsr"#;

#[test]
fn test_get_hsr_iface_yaml() {
    with_hsr_iface(|| {
        let state = NetState::retrieve().unwrap();
        let iface = &state.ifaces[IFACE_NAME];

        assert_eq!(iface.iface_type, crate::IfaceType::Hsr);

        assert_value_match(EXPECTED_HSR_INFO, iface);
    });
}

#[test]
#[ignore]
fn test_get_hsr_interlink_iface_yaml() {
    with_hsr_interlink_iface(|| {
        let state = NetState::retrieve().unwrap();
        let iface = &state.ifaces[IFACE_NAME];

        assert_eq!(iface.iface_type, crate::IfaceType::Hsr);

        assert_value_match(EXPECTED_HSR_INTERLINK_INFO, iface);
    });
}

fn with_hsr_iface<T>(test: T)
where
    T: FnOnce() + panic::UnwindSafe,
{
    super::utils::set_network_environment("hsr");

    let result = panic::catch_unwind(|| {
        test();
    });

    super::utils::clear_network_environment();
    assert!(result.is_ok())
}

fn with_hsr_interlink_iface<T>(test: T)
where
    T: FnOnce() + panic::UnwindSafe,
{
    super::utils::set_network_environment("hsr_interlink");

    let result = panic::catch_unwind(|| {
        test();
    });

    super::utils::clear_network_environment();
    assert!(result.is_ok())
}
